gloo/allreduce.cc

```diff
@@ -131,14 +132,23 @@ void allreduce(const detail::AllreduceOptionsImpl& opts) {
     return;
   }
 
-  switch (opts.algorithm) {
+  auto algorithm = opts.algorithm;
+  if (is_intra_node(opts)) {
```

Comment:
Doing a broadcast prior to every single allreduce seems pretty expensive -- can we move this logic to the TCP init in `createAndConnectAllPairs`? We have enough information there I believe to compute the topology since we have all the hostnames.

Reply:
The other option would be to inspect the pairs that are participating and check if they all share the same IP